### ISSUE TITLE
Discourage "color: transparent" for "background-clip: text" 

### ISSUE THREAD

**Issue by myfonj** [mdn/content#2638]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/background-clip

#### TL;DR:
- `color: transparent` is dangerous, maybe mention feature detection or propose text-fill-color instead;
- where is `background-image:`, there should be `background-color:`;
- especially when clipped into text.

#### What information was incorrect, unhelpful, or incomplete?

Code snippets use `color: transparent` or nearly transparent black for demonstrations of `background-clip: text`, with no support detection or other practice to preserve backward compatibility and fail-safe accessibility.  It sounds reasonable for demonstrative purposes, but is dangerous for UAs without proper support and for scenarios when decorative background image rendering fails for any reason.  MDN should at least warn about such things, or better provide best-practice solution.

#### Specific section or headline?

Intro and Examples.

#### What did you expect to see?

`color: transparent`  substituted with `-webkit-text-fill-color: transparent` or even better wrapped in `@supports` block in snippets. Plus concrete `color` and `background-color` fallback declarations. Or warning.

I know situation is complicated because there is no standard un-prefixed `text-fill-color` ATM but i guess all browsers supporting `-webkit-background-clip: text` supports `-webkit-text-fill-color`. (From compat tables stands out only Safari 3 (year ~2009) not supporting ..text-fill-color but supporting ..background-clip: text; cannot verify and have no clue how to cope with this, since `@supports` apparently came to Safari in v.9 (~2016).)

The thing is, I am not even sure what the best practice would be here and where to put it in the page.  I tried to make truly resilient snippet preserving readability and contrast in UAs without support of modern CSS and with failing resources in mind, but resulting code looks quite bloated: https://codepen.io/myf/pen/zYoPJam?editors=1100, try to cripple rules to simulate failed resources or lack CSS of support:

```CSS

/* 
 - assumed dark text on bright background;
 - gradients in background-images are illustrative as "external resources that could fail or be blocked"
*/

/* § ancient / fallback / default
*/

p {
	background-color: pink;
	background-image: repeating-linear-gradient(white, pink 0.5em, white 1em);
	color: black;
}

/* § modern / fancy / extra
*/

@supports (-webkit-background-clip: text) and (-webkit-text-fill-color: transparent) {
	p {
		background-color: black;
		background-image: none;
		/* in case following image failed, previous reset is needed to negate the legacy bright one
		*/
		background-image: repeating-radial-gradient(black, darkred 0.5em, black 1em);
		color: white;
		-webkit-background-clip: text;
		-webkit-text-fill-color: transparent;
		/* color: transparent should be safe here as well,
		 but I just cannot supress all those red flags seeing `color: transparent` on important content
		*/
	}
}

```

Sorry for half-baked solution and murky guesswork about actual best practices.  As demonstrated, I couldn't propose consistent solution, but I really had to vent this itch.

#### Did you test this? If so, how?

From compat data It is clearly apparent that current examples would produce invisible text in IE.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/css/background-clip`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/background-clip
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/background-clip/index.html
* Last commit: https://github.com/mdn/content/commit/2a14c0eac061493b1160076a3e72ec8e346d2223
* Document last modified: 2021-02-19T18:56:36.000Z

</details>

**Comment by rachelandrew** [comment]
@ericwbailey do you have any thoughts on this? Seems a potential a11y issue with regard to contrast.

**Comment by ericwbailey** [comment]
I tested [this pen](https://codepen.io/ericwbailey/pen/LYbejwK) with NVDA/Firefox. It was able to identify/read out the heading correctly, same `p` and other text elements. This is not a comprehensive test, but the Accessibility Inspector seems to be reporting it correctly.

The other issues I could see with this from an a11y perspective:

- The background image does not grant sufficient contrast to pass [1.4.3](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html).
- The background image fails to load, which props up @myfonj's suggestion for including a fallback background color.

That being said, support for `background-clip` is good for evergreen browsers. This feels like it could be a short note about `@supports` to me.

**Comment by myfonj** [comment]
Good to have verified `color: transparent` does not affect screen readers.  I didn't expect there will be problem there -- it's just a colour anyway -- but nice to be sure nonetheless.

Anyway, I see I expressed my concerns really bad and mangled several loosely related aspects into single undecipherable report, sorry for that.  To untangle:

## First (main) concern: Graceful degradation for *non-evergreen* UAs

(I suppose you already know where I am heading, and I feel a silly to 'mansplain' it b/c I know whom I am talking to, but for completeness here I go.) There was about twenty years long period of **UAs supporting `color: transparent`** before `background-clip` with `text` value became widely supported (say, around 2017).  Nature of CSS is usually introducing new features on top of old so that non-conforming UAs ignore them without losing usability: e.g. feeding `div { display: grid; }` to any old UA will just "do nothing" and div's will keep their boring `block` nature but will remain accessible and readable.  This `background-clip: text` is special, because it works only when accompanied with transparent text, and that combination is what I called "dangerous" for this very reason: **it produces invisible text in any UA with CSS support older than four years or so.**  And that's what I feel is quite against the nature of web platform, and that MDN should not encourage non-knowledgeable reuse and spread of code snippets that might damage use experience of minor groups of users with 'obscure' UAs.  And that's why I feel much more confident using `text-fill-color: transparent` instead of `color: transparent`, because as it seems support of both properties is mostly aligned across UAs (they either support both or none, with mentioned caveat about ancient Safari and lack of proper testing on my side).

---

Eric managed to pull out the remaining two other concerns.  So just repeating here: 

## Second concern: Unfulfilled expectations about image rendering

Images in CSS are most often decorative and presentations should not rely on them when conveying information.  Also they may be blocked, broken or not loaded.  Such negative outcome is mostly preventable following standard recommendation of setting `background-color` along `color`, so yes, repeating it in snippets or noting it along 

## Third concern: Visual contrast of clipped images

I admit strictly speaking this have nothing to do with `color: transparent` per se since this is about "everything works fine" scenario.  My line of thought is that good "light decorative image under dark text" cannot by definition be used as "dark fill of text on bright background", and even "dark decorative image under bright text" may have insufficient contrast when used as "dark fill on bright background".  And as (insufficiently) demonstrated in my snippet, when previous concerns creeps in it makes whole problem lot more complicated.

---

I hope it is more clear now, and sorry again for wall of text in really bad English.  If time permits and necessity arises I'll dig more into this problem, maybe make concise overview of all scenarios, gather concrete data and hopefully come up with terse and clear solution.  (Maybe it is just unnecessary scaremongering from my side after all…)

**Comment by Ryuno-Ki** [comment]
@myfonj Your English was not bad at all (but I'm not a native speaker as well).
In fact, I learned something new from it! :rocket:
